File: C#/Aplicatii client-server/labClientServerCSharp/ClientServer/Networking/ClientObjectWorker.cs
using lab3.domain;
using Networking.dto;
using Services;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace Networking
{
    public class ClientWorker: IObserver
    {
        private IServices server;
        private TcpClient connection;

        private NetworkStream stream;
        private IFormatter formatter;
        private volatile bool connected;
        public ClientWorker(IServices server, TcpClient connection)
        {
            this.server = server;
            this.connection = connection;
            try
            {
                stream = connection.GetStream();
                formatter = new BinaryFormatter();
                connected = true;
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        public virtual void run()
        {
            while (connected)
            {
                try
                {
                    object request = formatter.Deserialize(stream);
                    object response = handleRequest((Request)request);
                    if (response != null)
                    {
                        sendResponse((Response)response);
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }

                try
                {
                    Thread.Sleep(1000);
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }
            }
            try
            {
                stream.Close();
                connection.Close();
            }
            catch (Exception e)
            {
                Console.WriteLine("Error " + e);
            }
        }

        private Response handleRequest(Request request)
        {
            Response response = null;
            if(request is LoginRequest)
            {
                Console.WriteLine("Login request...");
                LoginRequest loginRequest = (LoginRequest)request;
                OficiuDTO oficiuDTO = loginRequest.OficiuDTO;
                Oficiu oficiu = DTOUtils.getFromDTO(oficiuDTO);
                try
                {
                    lock(server)
                    {
                        server.login(oficiu, this);
                    }
                    return new OkResponse();
                }
                catch(MyException ex)
                {
                    connected = false;
                    return new ErrorResponse(ex.Message);
                }
            }

            if(request is LogoutRequest)
            {
                Console.WriteLine("Logout request...");
                LogoutRequest logoutRequest = (LogoutRequest)request;
                OficiuDTO oficiuDTO = logoutRequest.OficiuDTO;
                Oficiu oficiu = DTOUtils.getFromDTO(oficiuDTO);
                try
                {
                    lock (server)
                    {
                        server.logout(oficiu, this);
                    }
                    connected = false;
                    return new OkResponse();
                }
                catch(MyException ex)
                {
                    return new ErrorResponse(ex.Message);
                }
            }

            if (request is FindOficiuByUserPassRequest)
            {
                Console.WriteLine("FindOficiuByUserPass request...");
                FindOficiuByUserPassRequest findRequest = (FindOficiuByUserPassRequest)request;
                string username = findRequest.Username;
                string password = findRequest.Password;
                try
                {
                    lock (server)
                    {
                        Oficiu oficiu = server.findOficiuByUserPass(username, password);
                        if (oficiu != null)
                        {
                            OficiuDTO oficiuDTO = DTOUtils.getDTO(oficiu);
                            return new FindOficiuByUserPassResponse(oficiuDTO);
                        }
                        else
                        {
                            connected = false;
                            return new ErrorResponse("Username si parola gresite!");
                        }
                    }
                }
                catch(MyException ex)
                {
                    connected = false;
                    return new ErrorResponse("Username si parola gresite!");
                }
            }

            if(request is FindAllCursaRequest)
            {
                Console.WriteLine("FindAllCursa request...");
                try
                {
                    lock (server)
                    {
                        Cursa[] curse = server.findAllCursa();
                        CursaDTO[] cursaDTOs = DTOUtils.getDTO(curse);
                        return new FindAllCursaResponse(cursaDTOs);
                    }
                }
                catch(MyException ex)
                {
                    return new ErrorResponse(ex.Message);
                }
            }

            if(request is FindAllParticipantRequest)
            {
                Console.WriteLine("FindAllParticipant request...");
                try
                {
                    lock (server)
                    {
                        Participant[] participanti = server.findAllParticipant();
                        ParticipantDTO[] participantDTOs = DTOUtils.getDTO(participanti);
                        return new FindAllParticipantResponse(participantDTOs);
                    }
                }
                catch (MyException ex)
                {
                    return new ErrorResponse(ex.Message);
                }
            }

            if (request is FindAllParticipantEchipaRequest)
            {
                Console.WriteLine("FindAllParticipantEchipa request...");
                FindAllParticipantEchipaRequest findRequest = (FindAllParticipantEchipaRequest)request;
                string echipa = findRequest.Echipa;
                try
                {
                    lock (server)
                    {
                        Participant[] participanti = server.findAllParticipant(echipa);
                        ParticipantDTO[] participantDTOs = DTOUtils.getDTO(participanti);
                        return new FindAllParticipantEchipaResponse(participantDTOs);
                    }
                }
                catch (MyException ex)
                {
                    return new ErrorResponse(ex.Message);
                }
            }

            if(request is FindParticipantRequest)
            {
                Console.WriteLine("FindParticipant request...");
                FindParticipantRequest findRequest = (FindParticipantRequest)request;
                string nume = findRequest.Nume;
                string echipa = findRequest.Echipa;
                int capMotor = findRequest.CapMotor;
                try
                {
                    lock (server)
                    {
                        Participant participant = server.findParticipant(nume, echipa, capMotor);
                        ParticipantDTO participantDTO = DTOUtils.getDTO(participant);
                        return new FindParticipantResponse(participantDTO);
                    }
                }
                catch (MyException ex)
                {
                    return new ErrorResponse(ex.Message);
                }
            }

            if(request is FindCursaRequest)
            {
                Console.WriteLine("FindCursa request...");
                FindCursaRequest findRequest = (FindCursaRequest)request;
                string id = findRequest.Id;
                try
                {
                    lock (server)
                    {
                        Cursa cursa = server.findCursa(id);
                        CursaDTO cursaDTO = DTOUtils.getDTO(cursa);
                        return new FindCursaResponse(cursaDTO);
                    }
                }
                catch (MyException ex)
                {
                    return new ErrorResponse(ex.Message);
                }
            }

            if(request is SaveParticipantCursaRequest)
            {
                Console.WriteLine("SaveParticipantCursa request...");
                SaveParticipantCursaRequest saveRequest = (SaveParticipantCursaRequest)request;
                ParticipantCursaDTO participantCursaDTO = saveRequest.ParticipantCursaDTO;
                ParticipantCursa participantCursa = DTOUtils.getFromDTO(participantCursaDTO);
                try
                {
                    lock (server)
                    {
                        server.saveParticipantCursa(participantCursa);
                    }
                    return new OkResponse();
                }
                catch (MyException ex)
                {
                    return new ErrorResponse(ex.Message);
                }
            }

            if(request is DeleteCursaRequest)
            {
                Console.WriteLine("DeleteCursa request...");
                DeleteCursaRequest deleteRequest = (DeleteCursaRequest)request;
                string id = deleteRequest.Id;
                try
                {
                    lock (server)
                    {
                        server.deleteCursa(id);
                    }
                    return new OkResponse();
                }
                catch(MyException ex)
                {
                    return new ErrorResponse(ex.Message);
                }
            }

            if(request is SaveCursaRequest)
            {
                Console.WriteLine("SaveCursa request...");
                SaveCursaRequest saveRequest = (SaveCursaRequest)request;
                CursaDTO cursaDTO = saveRequest.CursaDTO;
                Cursa cursa = DTOUtils.getFromDTO(cursaDTO);
                try
                {
                    lock (server)
                    {
                        server.saveCursa(cursa);
                    }
                    return new OkResponse();
                }
                catch (MyException ex)
                {
                    return new ErrorResponse(ex.Message);
                }
            }

            if (request is NotifyRequest)
            {
                Console.WriteLine("Notify request...");
                NotifyRequest notifyRequest = (NotifyRequest)request;
                CursaDTO[] cursaDTOs = notifyRequest.CursaDTOs;
                Cursa[] curse = DTOUtils.getFromDTO(cursaDTOs);
                try
                {
                    lock (server)
                    {
                        server.submitted(curse);
                    }
                    return new OkResponse();
                }
                catch (MyException ex)
                {
                    return new ErrorResponse(ex.Message);
                }
            }

            return response;
        }

        private void sendResponse(Response response)
        {
            Console.WriteLine("sending response " + response);
            formatter.Serialize(stream, response);
            stream.Flush();
        }

        public virtual void participantCursaAdded(Cursa[] curse)
        {
            Console.WriteLine("New participantCursa added...");
            CursaDTO[] cursaDTOs = DTOUtils.getDTO(curse);
            try
            {
                sendResponse(new NewParticipantCursaResponse(cursaDTOs));
            }
            catch (Exception e)
            {
                throw new MyException("Sending error " + e);
            }
        }
    }
}

File: Java/Aplicatii client-server/labJavaClientServerServiciiRestReactJS/Networking/src/main/java/objectProtocol/ClientObjectWorker.java
package objectProtocol;

import dto.*;
import model.Cursa;
import model.Oficiu;
import model.Participant;
import model.ParticipantCursa;
import services.IObserver;
import services.IServices;
import services.MyException;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;

public class ClientObjectWorker implements Runnable, IObserver {
    private IServices server;
    private Socket connection;

    private ObjectInputStream input;
    private ObjectOutputStream output;
    private volatile boolean connected;

    public ClientObjectWorker(IServices server, Socket connection){
        this.server = server;
        this.connection = connection;
        try{
            output = new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input = new ObjectInputStream(connection.getInputStream());
            connected = true;
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
    @Override
    public void run() {
        while(connected){
            try{
                Object request = input.readObject();
                Object response = handleRequest((Request) request);
                if(response != null){
                    sendResponse((Response) response);
                }
            } catch (IOException | ClassNotFoundException | MyException e) {
                e.printStackTrace();
            }
            try{
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        try{
            input.close();
            output.close();
            connection.close();
        } catch (IOException e) {
            System.out.println("Error " + e);
        }
    }

    public void participantCursaAdded() throws MyException{
        System.out.println("ParticipantCursa added");
        try{
            sendResponse(new NewParticipantCursaAddedResponse());
        }
        catch (IOException e){
            throw new MyException("Sending error: " + e);
        }
    }

    private void sendResponse(Response response) throws IOException{
        System.out.println("Sending response..." + response);
        output.writeObject(response);
        output.flush();
    }

    private Response handleRequest(Request request) throws MyException {
        Response response = null;
        if(request instanceof LoginRequest){
            System.out.println("Login request...");
            LoginRequest loginRequest = (LoginRequest) request;
            OficiuDTO oficiuDTO = loginRequest.getOficiuDTO();
            Oficiu oficiu = DTOUtils.getfromDTO(oficiuDTO);
            try{
                server.findOficiuByUserPass(oficiu.getUsername(), oficiu.getPassword(), this);
                return new OkResponse();
            } catch (MyException e) {
                connected = false;
                return new ErrorResponse(e.getMessage());
            }
        }

        if(request instanceof GetOficiuByUserPassRequest){
            System.out.println("GetOficiuByUserPass request");
            GetOficiuByUserPassRequest req = (GetOficiuByUserPassRequest) request;
            String username = req.getUsername();
            String password = req.getPassword();
            Oficiu oficiu = server.findOficiuByUserPass(username, password);
            OficiuDTO oficiuDTO = DTOUtils.getDTO(oficiu);
            return new GetOficiuByUserPassResponse(oficiuDTO);
        }

        if(request instanceof LogoutRequest){
            System.out.println("Logout request");
            LogoutRequest logoutRequest = (LogoutRequest) request;
            OficiuDTO oficiuDTO = logoutRequest.getOficiuDTO();
            Oficiu oficiu = DTOUtils.getfromDTO(oficiuDTO);
            System.out.println("Oficiul in ClientWorker este: " + oficiu);
            try {
                server.logout(oficiu, this);
                connected = false;
                return new OkResponse();
            } catch (MyException e) {
                return new ErrorResponse(e.getMessage());
            }
        }

        if(request instanceof GetAllParticipantsRequest){
            System.out.println("GetAllParticipants request");
            try{
                Participant[] participants = server.findAllParticipant();
                return new GetAllParticipantsResponse(participants);
            } catch (MyException e) {
                return new ErrorResponse(e.getMessage());
            }
        }

        if(request instanceof GetAllCursaRequest){
            System.out.println("GetAllCursa request");
            try{
                Cursa[] curse = server.findAllCursa();
                CursaDTO[] curseDTO = DTOUtils.getDTO(curse);
                return new GetAllCursaResponse(curseDTO);
            } catch (MyException e) {
                return new ErrorResponse(e.getMessage());
            }
        }

        if(request instanceof  GetAllParticipantsByTeamRequest){
            System.out.println("GetAllParticipantsByTeam request");
            GetAllParticipantsByTeamRequest req = (GetAllParticipantsByTeamRequest) request;
            String echipa = req.getEchipa();
            try{
                Participant[] participants = server.findAllParticipant(echipa);
                return new GetAllParticipantsResponse(participants);
            } catch (MyException e) {
                return new ErrorResponse(e.getMessage());
            }
        }

        if(request instanceof GetEchipeRequest){
            System.out.println("GetEchipe request");
            try{
                String echipe = server.findAllEchipe();
                return new GetEchipeResponse(echipe);
            } catch (MyException e) {
                return new ErrorResponse(e.getMessage());
            }
        }

        if(request instanceof GetParticipantRequest){
            System.out.println("GetParticipant request");
            GetParticipantRequest req = (GetParticipantRequest) request;
            String nume = req.getNume();
            String echipa = req.getEchipa();
            Integer capMotor = req.getCapMotor();
            try{
                Participant participant = server.findParticipant(nume, echipa, capMotor);
                ParticipantDTO participantDTO = DTOUtils.getDTO(participant);
                return new GetParticipantResponse(participantDTO);
            } catch (MyException e) {
                return new ErrorResponse(e.getMessage());
            }
        }

        if(request instanceof GetCursaRequest){
            System.out.println("GetCursa request");
            GetCursaRequest req = (GetCursaRequest) request;
            String idCursa = req.getIdCursa();
            try{
                Cursa cursa = server.findCursa(idCursa);
                return new GetCursaResponse(cursa);
            } catch (MyException e) {
                return new ErrorResponse(e.getMessage());
            }
        }

        if(request instanceof AddParticipantCursaRequest){
            System.out.println("AddParticipantCursa request");
            AddParticipantCursaRequest req = (AddParticipantCursaRequest) request;
            ParticipantCursaDTO participantCursaDTO = req.getParticipantCursaDTO();
            System.out.println("Dto: ");
            System.out.println(participantCursaDTO);
            ParticipantCursa participantCursa = DTOUtils.getFromDTO(participantCursaDTO);
            try{
                System.out.println("Participantul Cursa de salvat este:");
                System.out.println(participantCursa);
                server.saveParticipantCursa(participantCursa);
                return new OkResponse();
            } catch (MyException e) {
                return new ErrorResponse(e.getMessage());
            }
        }

        if(request instanceof UpdateCursaRequest){
            System.out.println("UpdateCursa request");
            UpdateCursaRequest req = (UpdateCursaRequest) request;
            String idCursa = req.getIdCursa();
            CursaDTO cursaDTO = req.getCursaDTO();
            Cursa cursa = DTOUtils.getfromDTO(cursaDTO);
            try{
                server.updateCursa(idCursa, cursa);
                return new OkResponse();
            } catch (MyException e) {
                return new ErrorResponse(e.getMessage());
            }
        }

        return response;
    }

}
